src/service.rs

```diff
@@ -45,7 +46,62 @@ mod ip_vote;
 mod query_info;
 mod test;
 
-use hashset_delay::HashSetDelay;
+/// Request type for Protocols using `TalkReq` message.
+///
+/// Automatically responds with an empty body on drop if
+/// [`TalkRequest::respond`] is not called.
+#[derive(Debug)]
+pub struct TalkRequest {
+    id: RequestId,
+    node_address: NodeAddress,
+    protocol: Vec<u8>,
+    body: Vec<u8>,
+    sender: Option<mpsc::UnboundedSender<HandlerRequest>>,
+}
+
+impl Drop for TalkRequest {
+    fn drop(&mut self) {
+        let sender = match self.sender.take() {
+            Some(s) => s,
+            None => return,
+        };
+
+        let response = Response {
+            id: self.id.clone(),
+            body: ResponseBody::Talk { response: vec![] },
+        };
+
+        debug!("Sending empty TALK response to {}", self.node_address);
+        let _ = sender.send(HandlerRequest::Response(
+            self.node_address.clone(),
+            Box::new(response),
+        ));
+    }
+}
+
+impl TalkRequest {
+    pub fn protocol(&self) -> &[u8] {
+        &self.protocol
+    }
+
+    pub fn body(&self) -> &[u8] {
+        &self.body
+    }
+
+    pub fn respond(mut self, response: Vec<u8>) {
+        debug!("Sending TALK response to {}", self.node_address);
+
+        let response = Response {
+            id: self.id.clone(),
+            body: ResponseBody::Talk { response },
+        };
+
+        let _ = self.sender.take().unwrap().send(HandlerRequest::Response(
```

Comment:
Errors can arise here in the case that we have shutdown but still attempt to send responses. So might be nice to allow the application to handle these.

Reply:
>  Also, if a user tries to call respond twice, their program will panic.

`self` is not taken by reference, so it can't be called twice as far as I know.

> Errors can arise here in the case that we have shutdown but still attempt to send responses. So might be nice to allow the application to handle these.

Agreed! I just followed the [original implementation](https://github.com/sigp/discv5/blob/19ede3cb8dec95b081e59dc9275128a3243fee4f/src/service.rs#L491), but would be happy to return some error. It looks like most of the other `send` calls either ignore the error, or log something. Is there an error type I should use and/or extend?

Reply:
Oh yeah, sorry missed that this takes ownership of self.

I haven't really exposed specific responses to the user before. I guess we could use a `RequestError`: https://github.com/sigp/discv5/blob/master/src/error.rs#L73 

Or if you think appropriate could make our own ResponseError.

Reply:
For some of the internal messaging that's not naturally exposed to the user, I have been throwing away some errors in these circumstances. I should at least log some of the errors if the channel fails. 

Just a bit more conscious of this one as it will be directly exposed to the users.